hyperdome_server/web.py

```diff
@@ -0,0 +1,407 @@
+        def counselor_signin():
+            counselor_key = request.form["pub_key"]
+            # TODO authenticate
+            # user = load_user(request.form['username'])
+            sid = self.common.random_string(16)
+            # will use capacity variable for this later
+            self.counselors_available[sid] = 1
+            self.counselor_keys[sid] = counselor_key
+            return sid
+
+        @self.app.route("/generate_guest_id")
+        def generate_guest_id():
+            # TODO check for collisions
+            return self.common.random_string(16)
+
+        @self.app.route("/send_message", methods=["POST"])
+        def message_from_user():
+            message = request.form["message"]
+            user_id = request.form["user_id"]
+            if user_id not in self.active_chat_user_map:
+                return "no chat", 404
+            other_user = self.active_chat_user_map[user_id]
+            if other_user in self.pending_messages:
+                self.pending_messages[other_user] += f"\n{message}"
+            elif other_user:  # may be empty string if other disconnected
+                self.pending_messages[other_user] = message
+            else:
+                return "user left", 404
+            return "Success"
+
+        @self.app.route("/chat_status")
+        def chat_status():
+            user_id = request.form["user_id"]
+            try:
+                return (
+                    "CHAT_OVER"
+                    if self.active_chat_user_map[user_id] == ""
+                    else "CHAT_ACTIVE"
+                )
+            except KeyError:
+                return "NO_CHAT"
+
+        @self.app.route("/collect_messages", methods=["GET"])
+        def collect_messages():
+            guest_id = request.form["user_id"]
+            return self.pending_messages.pop(guest_id, "")
+
+    def error404(self):
+        self.add_request(Web.REQUEST_OTHER, request.path)
+        if request.path != "/favicon.ico":
+            self.error404_count += 1
+
+            # In receive mode, with public mode enabled, skip rate limiting
+            # 404s
+            if self.error404_count == 20:
+                self.add_request(Web.REQUEST_RATE_LIMIT, request.path)
+                self.force_shutdown()
+
+        r = make_response(render_template("404.html"), 404)
+        return self.add_security_headers(r)
+
+    def error403(self):
+        self.add_request(Web.REQUEST_OTHER, request.path)
+
+        r = make_response(render_template("403.html"), 403)
+        return self.add_security_headers(r)
+
+    def add_security_headers(self, r):
+        """
+        Add security headers to a request
+        """
+        for header, value in self.security_headers:
+            r.headers.set(header, value)
+        return r
+
+    def _safe_select_jinja_autoescape(self, filename):
+        return filename is None or filename.endswith(
+            (".html", ".htm", ".xml", ".xhtml")
```

Comment:
I didn't know endswith could do multiple options, I can use that a lot :o

Reply:
huh, they don't even really show it in documentation either, it's just casually mentioned "oh you can use a tuple too" without really noting the implications.